controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -271,60 +239,29 @@ private String getTransactionName(String scope, String stream, long segmentId, U
-            public void connectionDropped() {
-                log.warn("commitTransaction {} connection dropped", transactionName);
-                result.completeExceptionally(
-                        new WireCommandFailedException(type, WireCommandFailedException.Reason.ConnectionDropped));
-            }
-
-            @Override
-            public void wrongHost(WireCommands.WrongHost wrongHost) {
-                log.warn("commitTransaction {} wrongHost", transactionName);
-                result.completeExceptionally(
-                        new WireCommandFailedException(type, WireCommandFailedException.Reason.UnknownHost));
-            }
 
-            @Override
-            public void segmentsMerged(WireCommands.SegmentsMerged segmentsMerged) {
-                log.debug("commitTransaction {} TransactionCommitted", transactionName);
-                result.complete(TxnStatus.newBuilder().setStatus(TxnStatus.Status.SUCCESS).build());
-            }
-
-            @Override
-            public void noSuchSegment(WireCommands.NoSuchSegment noSuchSegment) {
-                if (noSuchSegment.getSegment().equals(transactionName)) {
-                    log.info("commitTransaction {} NoSuchSegment", transactionName);
-                    result.complete(TxnStatus.newBuilder().setStatus(TxnStatus.Status.SUCCESS).build());
-                } else {
-                    log.warn("commitTransaction {} Source Segment not found", noSuchSegment.getSegment());
-                    result.complete(TxnStatus.newBuilder().setStatus(TxnStatus.Status.FAILURE).build());
-                }
-            }
-
-            @Override
-            public void processingFailure(Exception error) {
-                log.error("commitTransaction {} failed", transactionName, error);
-                handleError(error, result, type);
-            }
+        RawClient connection = new RawClient(ModelHelper.encode(uri), connectionFactory);
+        final long requestId = connection.getFlow().asLong();
+        
+        WireCommands.MergeSegments request = new WireCommands.MergeSegments(requestId,
+                qualifiedNameTarget, transactionName, delegationToken);
 
-            @Override
-            public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenCheckFailed) {
-                result.completeExceptionally(
-                        new WireCommandFailedException(new AuthenticationException(authTokenCheckFailed.toString()),
-                                type, WireCommandFailedException.Reason.AuthFailed)
-                );
-            }
-        };
+        return sendRequest(connection, requestId, request)
+                .thenApply(r -> {
+                    handleReply(requestId, r, connection, transactionName, WireCommands.MergeSegments.class, type);
+                    if (r instanceof WireCommands.NoSuchSegment) {
+                        if (((WireCommands.NoSuchSegment) r).getSegment().equals(transactionName)) {
```

Comment:
I suggest putting `(WireCommands.NoSuchSegment) r` into a local variable since you use it more than once.